src/language_server_protocol.rs

```diff
@@ -1829,41 +1829,72 @@ impl LanguageClient {
             return Ok(Value::Null);
         }
 
-        info!("Begin {}", lsp::request::CodeLensRequest::METHOD);
         let filename = self.vim()?.get_filename(params)?;
         let language_id = self.vim()?.get_languageId(&filename, params)?;
-        let client = self.get_client(&Some(language_id))?;
-        let input = lsp::CodeLensParams {
-            text_document: TextDocumentIdentifier {
-                uri: filename.to_url()?,
-            },
-        };
+        let capabilities = self.get(|state| state.capabilities.clone())?;
+        if let Some(initialize_result) = capabilities.get(&language_id) {
+            // XXX: the capabilities state field stores the initialize result, not the capabilities
+            // themselves, so we need to deserialize to InitializeResult.
+            let initialize_result: InitializeResult =
+                serde_json::from_value(initialize_result.clone())?;
+            let capabilities = initialize_result.capabilities;
+            if capabilities.code_lens_provider.is_some() {
+                info!("Begin {}", lsp::request::CodeLensRequest::METHOD);
+                let client = self.get_client(&Some(language_id))?;
+                let input = lsp::CodeLensParams {
+                    text_document: TextDocumentIdentifier {
+                        uri: filename.to_url()?,
+                    },
+                };
 
-        let results: Value = client.call(lsp::request::CodeLensRequest::METHOD, &input)?;
+                let results: Value = client.call(lsp::request::CodeLensRequest::METHOD, &input)?;
+                let code_lens: Option<Vec<CodeLens>> = serde_json::from_value(results.clone())?;
 
-        let code_lens: Option<Vec<CodeLens>> = serde_json::from_value(results.clone())?;
-        let mut resolved_code_lens = vec![];
-        if let Some(code_lens) = code_lens {
-            for item in code_lens {
-                let mut item = item;
-                if let Some(_d) = &item.data {
-                    if let Some(cl) = client.call(lsp::request::CodeLensResolve::METHOD, &item)? {
-                        item = cl;
+                if capabilities
+                    .code_lens_provider
+                    .unwrap()
```

Comment:
Nit: I would avoid using `unwrap` whenever possible except for prototype as it's going to quit the whole process if unwrapping on none.

Anyway, I can merge this in and update myself.